src/EditorFeatures/Core/Implementation/CodeDefinitionWindow/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,240 @@
+            {
+                _subscribedViews.Add(textView);
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
+                QueueUpdateForCaretPosition(textView.Caret.Position);
+            }
+        }
+
+        void ITextViewConnectionListener.SubjectBuffersDisconnected(ITextView textView, ConnectionReason reason, IReadOnlyCollection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
+            {
+                if (_subscribedViews.Contains(textView))
+                {
+                    _subscribedViews.Remove(textView);
+                    textView.Caret.PositionChanged -= OnTextViewCaretPositionChanged;
+                }
+            }
+        }
+
+        private void OnTextViewCaretPositionChanged(object? sender, CaretPositionChangedEventArgs e)
+        {
+            QueueUpdateForCaretPosition(e.NewPosition);
+        }
+
+        private void QueueUpdateForCaretPosition(CaretPosition caretPosition)
+        {
+            Contract.ThrowIfFalse(_threadingContext.JoinableTaskContext.IsOnMainThread);
+
+            // Cancel any pending update for this view
+            _currentUpdateCancellationToken?.Cancel();
+
+            // See if we moved somewhere else in a projection that we care about
+            var pointInRoslynSnapshot = caretPosition.Point.GetPoint(tb => tb.ContentType.IsOfType(ContentTypeNames.RoslynContentType), caretPosition.Affinity);
+            if (pointInRoslynSnapshot == null)
+            {
+                return;
+            }
+
+            _currentUpdateCancellationToken = new CancellationTokenSource();
+
+            var cancellationToken = _currentUpdateCancellationToken.Token;
+            var asyncToken = _asyncListener.BeginAsyncOperation(nameof(DefinitionContextTracker) + "." + nameof(QueueUpdateForCaretPosition));
+            UpdateForCaretPositionAsync(pointInRoslynSnapshot.Value, cancellationToken).CompletesAsyncOperation(asyncToken);
+        }
+
+        private async Task UpdateForCaretPositionAsync(SnapshotPoint pointInRoslynSnapshot, CancellationToken cancellationToken)
+        {
+            var locations = await GetContextFromPointAfterDelayAsync(pointInRoslynSnapshot, cancellationToken).ConfigureAwait(true);
+
+            await _codeDefinitionWindowService.SetContextAsync(locations, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<ImmutableArray<CodeDefinitionWindowLocation>> GetContextFromPointAfterDelayAsync(
+            SnapshotPoint pointInRoslynSnapshot, CancellationToken cancellationToken)
+        {
+            // TODO: Does this allocate too many tasks - should we switch to a queue like the classifier uses?
```

Comment:
yes.  switch to queue please.  it already handles all the batching logic.

Reply:
(see comments elsewhere that I'm not sure we have a useful utility here, or at least one that allocates any fewer tasks. This is an old comment of @Pilchie's and I'm not sure it applies anymore.)

Reply:
Resolving this and going to use the other conversation as primary.